### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Warhammer4000**
## Summary by CodeRabbit

- **New Features**
  - Implemented a new automated process for updating and tagging Unity package versions upon code pushes.
  - Added a workflow to generate and deploy documentation automatically for Unity projects.

- **Documentation**
  - Updated `.gitignore` to better suit Unity and JetBrains Rider IDE.
  - Introduced a `.vsconfig` file to specify Visual Studio configurations for managed game development.

- **Enhancements**
  - Expanded the "Dialogue System" package with new classes for dialogue management, node-based system, editor tools, and UI management.
  - Updated Unity project settings to improve serialization, build settings, editor preferences, and package management.

- **Bug Fixes**
  - Adjusted Unity project settings to fix issues with graphics, physics, and memory configurations.

- **Refactor**
  - Organized IDE-specific settings and version control system mappings for better project structure.

- **Style**
  - Updated TextMesh Pro assets for improved font appearance and compliance with font licenses.

### CODE REVIEW

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/Data/DialogueBaseNodeEditor.cs
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/Data/DialogueBaseNodeEditor.cs]: Incorrect naming convention.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Samples~/SubtitleDemo/Configuration/SubtitileSettings.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Samples~/SubtitleDemo/Configuration/SubtitileSettings.cs]: Naming convention for public vars not followed

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/UITabs/DialogueGraphEditorUI.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/UITabs/DialogueGraphEditorUI.cs]: extra space

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterData.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterData.cs:21]: We will likely have a finite set of reactions so enums would be less error prone.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs:37]: Public var naming conventions not followed

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs:33]: This is doing a lot of work including creating a hashset object everytime IsValid is called. It should be a method CheckValidity() instead of a property that hides the perf cost.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueBase.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueBase.cs]: Extra space.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueBase.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueBase.cs]: If we use CharacterData scriptable then this should be a reference to that, not the CharacterID . Otherwise no point in replacing CharacterID with CharacterData scriptabl

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs]: Extra space throughout the file

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs]: 250 per node is hardcoded. Should be a variable in the graph scriptable.
This also assumes that nodes will only go in the right direction. This will not hold true for option nodes.
Assuming we represent option nodes with each option as children arranged vertically, we need offsets in the y axis.
This should be a function on the node class:
`node.getNewChildPosition()`
that is overridden for option nodes and normal dialogue nodes properly

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs]: Should be a method  on the node class to connect/add a child.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Samples~/SubtitleDemo/DialogueUI.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Samples~/SubtitleDemo/DialogueUI.cs]: Assuming we want to change subtitle settings in runtime via a setting menu, this needs a public accessor

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Samples~/SubtitleDemo/DialogueUI.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Samples~/SubtitleDemo/DialogueUI.cs]: Assuming we want to change subtitle settings in runtime via a setting menu, this needs to be public so that we can tell the dialogueUI to update.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueManager.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueManager.cs]: If this is meant to be the main way to set the DialogueGraph to play, naming it SetDialogueGraph would make it clearer

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs]: no context menu to create a dialogueGraph

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueNode.cs
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueNode.cs]: DialogueNodes created through the context menu are not connected to a dialogue graph.  Do we even need to allow that? We can only allow creation through the graph and avoid this issue.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs]: Assuming the `GetOutputPort` & `GetInputPort` functions are from XNode, however for someone glancing at this code without prior knowledge of XNode would not be able to understand the use case. 

Consider adding comments / `///<summary>` to improve code readability.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Samples~/SubtitleDemo/DialogueUI.cs
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Samples~/SubtitleDemo/DialogueUI.cs]: Subscribed event is never unsubscribed.

**Review by BDeshiDev — CHANGES_REQUESTED** [review]
There isn't a context menu to create a dialogue graph. I can only create dialogue nodes that have a null dialogue graph reference...
The samples folder is in the repo is named Samples~. Unity ignores the  folder. So I cannot even open the demo scene. Please follow the samples publication guide on loop.
image
renaming the samples folder allows running the sample. However, the running the dialoguegraph throws nullrefs. Please ensue working samples.

Hardcoding node offsets will certainly require rework for option nodes.

Regarding dialogue lines specific triggers, I don't think [serializereference] like approach is appropriate. Since the graph already uses scriptables, we can have a DialogueEventTriggerBase scriptable and have derived classes for HintSystem/objective system. etc as nodes in the graph. That way, by opening the graph, we can see which line fire which hint/objective etc.

There are along naming convention and spacing issues.
At the very least, the samples and graph creation issues has to be fixed before approval.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Samples~/SubtitleDemo/DialogueUI.cs
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Samples~/SubtitleDemo/DialogueUI.cs]: Explicit declaration of access modifier.

**Review by N-Evan — CHANGES_REQUESTED** [review]
Overall, the code needs many structural changes. 

- Incorrect naming 
- Unnecessary whitespaces
- Fix hard-coded values in side functions. Even if it's strictly necessary, maybe using a configuration script to modify these values from the outside should a better choice unless it's cemented that they will never change. 
- Some comments with the use cases would provide better readability. 

## Functional 

- Should have an event firing after each singular dialogue.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterData.cs
Comment (Warhammer4000) [reply Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterData.cs:21]: as it's public repo. I'd rather insist on keeping string instead of hardcoding enums. Although we could provide a static string class generation method perhaps but that seems a bit over kill

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Samples~/SubtitleDemo/DialogueUI.cs
Comment (Warhammer4000) [reply Assets/Packages/com.studio23.ss2.dialoguesystem/Samples~/SubtitleDemo/DialogueUI.cs]: Samples folder items are expected to be thrown away. Rewriten for each game. Not boilerplate

**Pull request closed without merging**
